In [125]:
import pandas as pd
import csv
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [126]:
data = pd.read_csv('Daily Time - Responses.csv')
data

Timestamp    Subject Start Time  End Time  RPE  Email Address  \
0    15/10/2024 18:33:08    Physics   16:02:00  17:00:00    2            NaN   
1    15/10/2024 21:59:04    Physics   18:25:00  20:30:00    4            NaN   
2    15/10/2024 21:59:00  Chemistry   11:23:00  11:55:00    3            NaN   
3    16/10/2024 14:02:36    Physics   10:30:00  11:30:00    2            NaN   
4    16/10/2024 17:50:31  Chemistry   16:30:00  17:50:00    3            NaN   
..                   ...        ...        ...       ...  ...            ...   
118  10/12/2024 23:06:45    Maths-I   18:00:00  19:00:00    3            NaN   
119  10/12/2024 23:07:38    Physics   21:00:00  23:05:00    2            NaN   
120  11/12/2024 18:17:31  Chemistry   16:37:00  17:45:00    3            NaN   
121  11/12/2024 21:39:14  Chemistry   11:22:00  12:40:00    2            NaN   
122  11/12/2024 21:40:00  Chemistry   15:20:00  17:00:00    3            NaN   

      Name        Date  
0    Nupur  15/10/2024  
1    Nupur  15/10/2024  
2    Swara  15/10/2024  
3    Swara  16/10/2024  
4      Jui  16/10/2024  
..     ...         ...  
118  Swara         NaN  
119  Swara         NaN  
120    Jui         NaN  
121  Swara         NaN  
122  Swara         NaN  

[123 rows x 8 columns]

In [127]:
 data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Timestamp      123 non-null    object 
 1   Subject        123 non-null    object 
 2   Start Time     123 non-null    object 
 3   End Time       123 non-null    object 
 4   RPE            123 non-null    int64  
 5   Email Address  0 non-null      float64
 6   Name           123 non-null    object 
 7   Date           117 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 7.8+ KB


In [128]:
import datetime
data['Timestamp'] = data['Timestamp'].apply(lambda x : pd.to_datetime(str(x)))    #initially converting to string and then to datetime
data['dates'] = data['Timestamp'].dt.date

data['time'] = data['Timestamp'].dt.time

C:\Users\Pranoti\AppData\Local\Temp\ipykernel_30212\3259131127.py:2: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data['Timestamp'] = data['Timestamp'].apply(lambda x : pd.to_datetime(str(x)))    #initially converting to string and then to datetime


In [129]:
data

Timestamp    Subject Start Time  End Time  RPE  Email Address  \
0   2024-10-15 18:33:08    Physics   16:02:00  17:00:00    2            NaN   
1   2024-10-15 21:59:04    Physics   18:25:00  20:30:00    4            NaN   
2   2024-10-15 21:59:00  Chemistry   11:23:00  11:55:00    3            NaN   
3   2024-10-16 14:02:36    Physics   10:30:00  11:30:00    2            NaN   
4   2024-10-16 17:50:31  Chemistry   16:30:00  17:50:00    3            NaN   
..                  ...        ...        ...       ...  ...            ...   
118 2024-10-12 23:06:45    Maths-I   18:00:00  19:00:00    3            NaN   
119 2024-10-12 23:07:38    Physics   21:00:00  23:05:00    2            NaN   
120 2024-11-12 18:17:31  Chemistry   16:37:00  17:45:00    3            NaN   
121 2024-11-12 21:39:14  Chemistry   11:22:00  12:40:00    2            NaN   
122 2024-11-12 21:40:00  Chemistry   15:20:00  17:00:00    3            NaN   

      Name        Date       dates      time  
0    Nupur  15/10/2024  2024-10-15  18:33:08  
1    Nupur  15/10/2024  2024-10-15  21:59:04  
2    Swara  15/10/2024  2024-10-15  21:59:00  
3    Swara  16/10/2024  2024-10-16  14:02:36  
4      Jui  16/10/2024  2024-10-16  17:50:31  
..     ...         ...         ...       ...  
118  Swara         NaN  2024-10-12  23:06:45  
119  Swara         NaN  2024-10-12  23:07:38  
120    Jui         NaN  2024-11-12  18:17:31  
121  Swara         NaN  2024-11-12  21:39:14  
122  Swara         NaN  2024-11-12  21:40:00  

[123 rows x 10 columns]

In [130]:
df = data.drop(['Email Address', 'Timestamp', 'Date'], axis=1)

In [131]:
df

Subject Start Time  End Time  RPE   Name       dates      time
0      Physics   16:02:00  17:00:00    2  Nupur  2024-10-15  18:33:08
1      Physics   18:25:00  20:30:00    4  Nupur  2024-10-15  21:59:04
2    Chemistry   11:23:00  11:55:00    3  Swara  2024-10-15  21:59:00
3      Physics   10:30:00  11:30:00    2  Swara  2024-10-16  14:02:36
4    Chemistry   16:30:00  17:50:00    3    Jui  2024-10-16  17:50:31
..         ...        ...       ...  ...    ...         ...       ...
118    Maths-I   18:00:00  19:00:00    3  Swara  2024-10-12  23:06:45
119    Physics   21:00:00  23:05:00    2  Swara  2024-10-12  23:07:38
120  Chemistry   16:37:00  17:45:00    3    Jui  2024-11-12  18:17:31
121  Chemistry   11:22:00  12:40:00    2  Swara  2024-11-12  21:39:14
122  Chemistry   15:20:00  17:00:00    3  Swara  2024-11-12  21:40:00

[123 rows x 7 columns]

In [132]:
df['Start Time'] = df['Start Time'].apply(lambda x : pd.to_datetime(str(x)))
df['End Time'] = df['End Time'].apply(lambda x : pd.to_datetime(str(x)))

In [133]:
df

Subject          Start Time            End Time  RPE   Name  \
0      Physics 2025-01-15 16:02:00 2025-01-15 17:00:00    2  Nupur   
1      Physics 2025-01-15 18:25:00 2025-01-15 20:30:00    4  Nupur   
2    Chemistry 2025-01-15 11:23:00 2025-01-15 11:55:00    3  Swara   
3      Physics 2025-01-15 10:30:00 2025-01-15 11:30:00    2  Swara   
4    Chemistry 2025-01-15 16:30:00 2025-01-15 17:50:00    3    Jui   
..         ...                 ...                 ...  ...    ...   
118    Maths-I 2025-01-15 18:00:00 2025-01-15 19:00:00    3  Swara   
119    Physics 2025-01-15 21:00:00 2025-01-15 23:05:00    2  Swara   
120  Chemistry 2025-01-15 16:37:00 2025-01-15 17:45:00    3    Jui   
121  Chemistry 2025-01-15 11:22:00 2025-01-15 12:40:00    2  Swara   
122  Chemistry 2025-01-15 15:20:00 2025-01-15 17:00:00    3  Swara   

          dates      time  
0    2024-10-15  18:33:08  
1    2024-10-15  21:59:04  
2    2024-10-15  21:59:00  
3    2024-10-16  14:02:36  
4    2024-10-16  17:50:31  
..          ...       ...  
118  2024-10-12  23:06:45  
119  2024-10-12  23:07:38  
120  2024-11-12  18:17:31  
121  2024-11-12  21:39:14  
122  2024-11-12  21:40:00  

[123 rows x 7 columns]

In [134]:
df['Duration'] = df['End Time'] - df['Start Time']

In [135]:
df.drop(['Start Time', 'End Time'], axis=1)

Subject  RPE   Name       dates      time        Duration
0      Physics    2  Nupur  2024-10-15  18:33:08 0 days 00:58:00
1      Physics    4  Nupur  2024-10-15  21:59:04 0 days 02:05:00
2    Chemistry    3  Swara  2024-10-15  21:59:00 0 days 00:32:00
3      Physics    2  Swara  2024-10-16  14:02:36 0 days 01:00:00
4    Chemistry    3    Jui  2024-10-16  17:50:31 0 days 01:20:00
..         ...  ...    ...         ...       ...             ...
118    Maths-I    3  Swara  2024-10-12  23:06:45 0 days 01:00:00
119    Physics    2  Swara  2024-10-12  23:07:38 0 days 02:05:00
120  Chemistry    3    Jui  2024-11-12  18:17:31 0 days 01:08:00
121  Chemistry    2  Swara  2024-11-12  21:39:14 0 days 01:18:00
122  Chemistry    3  Swara  2024-11-12  21:40:00 0 days 01:40:00

[123 rows x 6 columns]

In [136]:
df['Duration'] = df['Duration'].astype(str).str.split('0 days ').str[-1]    #converting into string, seperating and then considering the time part
df

Subject          Start Time            End Time  RPE   Name  \
0      Physics 2025-01-15 16:02:00 2025-01-15 17:00:00    2  Nupur   
1      Physics 2025-01-15 18:25:00 2025-01-15 20:30:00    4  Nupur   
2    Chemistry 2025-01-15 11:23:00 2025-01-15 11:55:00    3  Swara   
3      Physics 2025-01-15 10:30:00 2025-01-15 11:30:00    2  Swara   
4    Chemistry 2025-01-15 16:30:00 2025-01-15 17:50:00    3    Jui   
..         ...                 ...                 ...  ...    ...   
118    Maths-I 2025-01-15 18:00:00 2025-01-15 19:00:00    3  Swara   
119    Physics 2025-01-15 21:00:00 2025-01-15 23:05:00    2  Swara   
120  Chemistry 2025-01-15 16:37:00 2025-01-15 17:45:00    3    Jui   
121  Chemistry 2025-01-15 11:22:00 2025-01-15 12:40:00    2  Swara   
122  Chemistry 2025-01-15 15:20:00 2025-01-15 17:00:00    3  Swara   

          dates      time  Duration  
0    2024-10-15  18:33:08  00:58:00  
1    2024-10-15  21:59:04  02:05:00  
2    2024-10-15  21:59:00  00:32:00  
3    2024-10-16  14:02:36  01:00:00  
4    2024-10-16  17:50:31  01:20:00  
..          ...       ...       ...  
118  2024-10-12  23:06:45  01:00:00  
119  2024-10-12  23:07:38  02:05:00  
120  2024-11-12  18:17:31  01:08:00  
121  2024-11-12  21:39:14  01:18:00  
122  2024-11-12  21:40:00  01:40:00  

[123 rows x 8 columns]

In [137]:
original_df = df.copy()

In [138]:
original_df

Subject          Start Time            End Time  RPE   Name  \
0      Physics 2025-01-15 16:02:00 2025-01-15 17:00:00    2  Nupur   
1      Physics 2025-01-15 18:25:00 2025-01-15 20:30:00    4  Nupur   
2    Chemistry 2025-01-15 11:23:00 2025-01-15 11:55:00    3  Swara   
3      Physics 2025-01-15 10:30:00 2025-01-15 11:30:00    2  Swara   
4    Chemistry 2025-01-15 16:30:00 2025-01-15 17:50:00    3    Jui   
..         ...                 ...                 ...  ...    ...   
118    Maths-I 2025-01-15 18:00:00 2025-01-15 19:00:00    3  Swara   
119    Physics 2025-01-15 21:00:00 2025-01-15 23:05:00    2  Swara   
120  Chemistry 2025-01-15 16:37:00 2025-01-15 17:45:00    3    Jui   
121  Chemistry 2025-01-15 11:22:00 2025-01-15 12:40:00    2  Swara   
122  Chemistry 2025-01-15 15:20:00 2025-01-15 17:00:00    3  Swara   

          dates      time  Duration  
0    2024-10-15  18:33:08  00:58:00  
1    2024-10-15  21:59:04  02:05:00  
2    2024-10-15  21:59:00  00:32:00  
3    2024-10-16  14:02:36  01:00:00  
4    2024-10-16  17:50:31  01:20:00  
..          ...       ...       ...  
118  2024-10-12  23:06:45  01:00:00  
119  2024-10-12  23:07:38  02:05:00  
120  2024-11-12  18:17:31  01:08:00  
121  2024-11-12  21:39:14  01:18:00  
122  2024-11-12  21:40:00  01:40:00  

[123 rows x 8 columns]

In [139]:
new_df = df.drop(['Start Time', 'End Time', 'dates', 'time', 'Subject', 'Duration'], axis =1)    #calculating the avg RPE of each student

In [140]:
new_df

RPE   Name
0      2  Nupur
1      4  Nupur
2      3  Swara
3      2  Swara
4      3    Jui
..   ...    ...
118    3  Swara
119    2  Swara
120    3    Jui
121    2  Swara
122    3  Swara

[123 rows x 2 columns]

In [141]:
new_df.groupby('Name').mean()

RPE
Name           
Jui    2.866667
Nupur  2.294118
Swara  2.161290

In [142]:
new_df1 = df.drop(['Start Time', 'End Time', 'dates', 'time', 'Subject', 'RPE'], axis =1)   #calculating the avg time allocated by each student daily
new_df1

Name  Duration
0    Nupur  00:58:00
1    Nupur  02:05:00
2    Swara  00:32:00
3    Swara  01:00:00
4      Jui  01:20:00
..     ...       ...
118  Swara  01:00:00
119  Swara  02:05:00
120    Jui  01:08:00
121  Swara  01:18:00
122  Swara  01:40:00

[123 rows x 2 columns]

In [143]:
new_df1

Name  Duration
0    Nupur  00:58:00
1    Nupur  02:05:00
2    Swara  00:32:00
3    Swara  01:00:00
4      Jui  01:20:00
..     ...       ...
118  Swara  01:00:00
119  Swara  02:05:00
120    Jui  01:08:00
121  Swara  01:18:00
122  Swara  01:40:00

[123 rows x 2 columns]

In [144]:
new_df1['Duration'] = df['Duration'].apply(lambda x : pd.to_datetime(str(x)))

In [145]:
new_df1['Hours'] = new_df1['Duration'].dt.hour
new_df1['Minutes'] = new_df1['Duration'].dt.minute

In [146]:
new_df1

Name            Duration  Hours  Minutes
0    Nupur 2025-01-15 00:58:00      0       58
1    Nupur 2025-01-15 02:05:00      2        5
2    Swara 2025-01-15 00:32:00      0       32
3    Swara 2025-01-15 01:00:00      1        0
4      Jui 2025-01-15 01:20:00      1       20
..     ...                 ...    ...      ...
118  Swara 2025-01-15 01:00:00      1        0
119  Swara 2025-01-15 02:05:00      2        5
120    Jui 2025-01-15 01:08:00      1        8
121  Swara 2025-01-15 01:18:00      1       18
122  Swara 2025-01-15 01:40:00      1       40

[123 rows x 4 columns]

In [147]:
new_df1 = new_df1.drop(['Duration'], axis=1)

In [148]:
new_df1

Name  Hours  Minutes
0    Nupur      0       58
1    Nupur      2        5
2    Swara      0       32
3    Swara      1        0
4      Jui      1       20
..     ...    ...      ...
118  Swara      1        0
119  Swara      2        5
120    Jui      1        8
121  Swara      1       18
122  Swara      1       40

[123 rows x 3 columns]

In [149]:
new_df1['Duration'] = new_df1['Minutes']+ new_df1['Hours']*60

In [150]:
new_df1

Name  Hours  Minutes  Duration
0    Nupur      0       58        58
1    Nupur      2        5       125
2    Swara      0       32        32
3    Swara      1        0        60
4      Jui      1       20        80
..     ...    ...      ...       ...
118  Swara      1        0        60
119  Swara      2        5       125
120    Jui      1        8        68
121  Swara      1       18        78
122  Swara      1       40       100

[123 rows x 4 columns]

In [151]:
new_df1 = new_df1.groupby('Name').mean()    #duration is in minutes

In [152]:
new_df1 = new_df1.drop(['Hours', 'Minutes'], axis=1)

In [153]:
new_df2 = original_df.drop(['Start Time', 'End Time', 'dates', 'time', 'Name', 'RPE'], axis =1)     #relation between subject and duration

In [155]:
new_df2

Subject  Duration
0      Physics  00:58:00
1      Physics  02:05:00
2    Chemistry  00:32:00
3      Physics  01:00:00
4    Chemistry  01:20:00
..         ...       ...
118    Maths-I  01:00:00
119    Physics  02:05:00
120  Chemistry  01:08:00
121  Chemistry  01:18:00
122  Chemistry  01:40:00

[123 rows x 2 columns]

In [156]:
new_df2['Duration'] = new_df2['Duration'].apply(lambda x : pd.to_datetime(str(x)))

In [157]:
new_df2['Hours'] = new_df2['Duration'].dt.hour
new_df2['Minutes'] = new_df2['Duration'].dt.minute

In [33]:
new_df2['Duration'] = new_df2['Minutes']+ new_df2['Hours']*60

In [34]:
new_df2

Subject  Duration  Hours  Minutes
0      Physics        58      0       58
1      Physics       125      2        5
2    Chemistry        32      0       32
3      Physics        60      1        0
4    Chemistry        80      1       20
..         ...       ...    ...      ...
118    Maths-I        60      1        0
119    Physics       125      2        5
120  Chemistry        68      1        8
121  Chemistry        78      1       18
122  Chemistry       100      1       40

[123 rows x 4 columns]

In [35]:
new_df2

Subject  Duration  Hours  Minutes
0      Physics        58      0       58
1      Physics       125      2        5
2    Chemistry        32      0       32
3      Physics        60      1        0
4    Chemistry        80      1       20
..         ...       ...    ...      ...
118    Maths-I        60      1        0
119    Physics       125      2        5
120  Chemistry        68      1        8
121  Chemistry        78      1       18
122  Chemistry       100      1       40

[123 rows x 4 columns]

In [36]:
new_df2 = new_df2.groupby('Subject').mean() 

In [37]:
new_df2.drop(['Hours', 'Minutes'], axis=1)

Duration
Subject              
Biology     70.000000
Chemistry   83.736842
English    102.428571
Maths-I     89.117647
Maths-II    83.736842
Physics     84.243243

In [38]:
diff_df = original_df[['Name', 'Subject', 'Start Time', 'End Time', 'RPE', 'Duration', 'time']]    #Nupur

In [39]:
diff_df = original_df[original_df['Name']== 'Nupur']

In [40]:
diff_df

Subject          Start Time            End Time  RPE   Name  \
0      Physics 2025-01-15 16:02:00 2025-01-15 17:00:00    2  Nupur   
1      Physics 2025-01-15 18:25:00 2025-01-15 20:30:00    4  Nupur   
10     Physics 2025-01-15 06:00:00 2025-01-15 08:30:00    3  Nupur   
11     Physics 2025-01-15 09:30:00 2025-01-15 11:45:00    4  Nupur   
12     English 2025-01-15 19:30:00 2025-01-15 21:35:00    2  Nupur   
13     English 2025-01-15 10:45:00 2025-01-15 12:00:00    1  Nupur   
27     Biology 2025-01-15 09:20:00 2025-01-15 10:45:00    2  Nupur   
28     Biology 2025-01-15 11:00:00 2025-01-15 12:00:00    1  Nupur   
49    Maths-II 2025-01-15 20:30:00 2025-01-15 21:30:00    2  Nupur   
50     Physics 2025-01-15 11:30:00 2025-01-15 12:00:00    1  Nupur   
51    Maths-II 2025-01-15 17:00:00 2025-01-15 18:10:00    3  Nupur   
62     Biology 2025-01-15 21:15:00 2025-01-15 22:45:00    1  Nupur   
94     Biology 2025-01-15 15:45:00 2025-01-15 16:30:00    2  Nupur   
95   Chemistry 2025-01-15 13:00:00 2025-01-15 14:00:00    2  Nupur   
100    Physics 2025-01-15 17:40:00 2025-01-15 19:00:00    3  Nupur   
108    Maths-I 2025-01-15 20:30:00 2025-01-15 21:50:00    3  Nupur   
112    Biology 2025-01-15 08:45:00 2025-01-15 09:55:00    3  Nupur   

          dates      time  Duration  
0    2024-10-15  18:33:08  00:58:00  
1    2024-10-15  21:59:04  02:05:00  
10   2024-10-17  19:45:51  02:30:00  
11   2024-10-17  19:47:16  02:15:00  
12   2024-10-17  19:48:33  02:05:00  
13   2024-10-17  19:49:27  01:15:00  
27   2024-10-18  12:07:23  01:25:00  
28   2024-10-18  12:07:49  01:00:00  
49   2024-10-22  17:58:29  01:00:00  
50   2024-10-22  17:58:59  00:30:00  
51   2024-10-22  18:19:58  01:10:00  
62   2024-10-26  18:37:57  01:30:00  
94   2024-11-22  16:33:57  00:45:00  
95   2024-11-25  20:58:16  01:00:00  
100  2024-11-26  19:01:18  01:20:00  
108  2024-11-28  22:15:45  01:20:00  
112  2024-11-30  09:57:34  01:10:00

In [41]:
diff_df1 = original_df[['Name', 'Subject', 'Start Time', 'End Time', 'RPE', 'Duration', 'time']]   #Swara

In [42]:
diff_df1 = original_df[original_df['Name']== 'Swara']

In [43]:
diff_df1

Subject          Start Time            End Time  RPE   Name  \
2    Chemistry 2025-01-15 11:23:00 2025-01-15 11:55:00    3  Swara   
3      Physics 2025-01-15 10:30:00 2025-01-15 11:30:00    2  Swara   
5    Chemistry 2025-01-15 12:00:00 2025-01-15 13:30:00    2  Swara   
6      Physics 2025-01-15 14:15:00 2025-01-15 15:15:00    3  Swara   
7    Chemistry 2025-01-15 07:15:00 2025-01-15 09:10:00    2  Swara   
14     English 2025-01-15 15:00:00 2025-01-15 17:00:00    2  Swara   
15     English 2025-01-15 20:00:00 2025-01-15 21:15:00    3  Swara   
16     English 2025-01-15 04:00:00 2025-01-15 06:07:00    1  Swara   
42     Physics 2025-01-15 19:00:00 2025-01-15 20:00:00    3  Swara   
45     Physics 2025-01-15 07:00:00 2025-01-15 09:00:00    2  Swara   
46   Chemistry 2025-01-15 09:15:00 2025-01-15 12:00:00    2  Swara   
47   Chemistry 2025-01-15 10:00:00 2025-01-15 11:00:00    1  Swara   
48   Chemistry 2025-01-15 16:00:00 2025-01-15 17:30:00    2  Swara   
52     Maths-I 2025-01-15 06:00:00 2025-01-15 09:00:00    1  Swara   
53    Maths-II 2025-01-15 10:00:00 2025-01-15 11:45:00    1  Swara   
63     Maths-I 2025-01-15 12:30:00 2025-01-15 13:30:00    2  Swara   
64     Physics 2025-01-15 14:00:00 2025-01-15 15:30:00    2  Swara   
65    Maths-II 2025-01-15 16:00:00 2025-01-15 17:00:00    2  Swara   
70    Maths-II 2025-01-15 12:30:00 2025-01-15 13:30:00    2  Swara   
71   Chemistry 2025-01-15 16:00:00 2025-01-15 18:00:00    2  Swara   
90   Chemistry 2025-01-15 08:45:00 2025-01-15 11:00:00    2  Swara   
96     Maths-I 2025-01-15 10:00:00 2025-01-15 14:00:00    2  Swara   
98     Physics 2025-01-15 08:33:00 2025-01-15 10:00:00    3  Swara   
99     Physics 2025-01-15 10:10:00 2025-01-15 12:55:00    2  Swara   
113    Physics 2025-01-15 10:00:00 2025-01-15 12:00:00    2  Swara   
116   Maths-II 2025-01-15 15:00:00 2025-01-15 18:00:00    2  Swara   
117   Maths-II 2025-01-15 16:20:00 2025-01-15 18:00:00    4  Swara   
118    Maths-I 2025-01-15 18:00:00 2025-01-15 19:00:00    3  Swara   
119    Physics 2025-01-15 21:00:00 2025-01-15 23:05:00    2  Swara   
121  Chemistry 2025-01-15 11:22:00 2025-01-15 12:40:00    2  Swara   
122  Chemistry 2025-01-15 15:20:00 2025-01-15 17:00:00    3  Swara   

          dates      time  Duration  
2    2024-10-15  21:59:00  00:32:00  
3    2024-10-16  14:02:36  01:00:00  
5    2024-10-16  22:10:18  01:30:00  
6    2024-10-16  22:11:01  01:00:00  
7    2024-10-17  09:41:25  01:55:00  
14   2024-10-17  21:26:07  02:00:00  
15   2024-10-17  21:27:34  01:15:00  
16   2024-10-18  06:10:10  02:07:00  
42   2024-10-18  22:36:46  01:00:00  
45   2024-10-22  06:33:24  02:00:00  
46   2024-10-22  06:34:09  02:45:00  
47   2024-10-22  06:34:47  01:00:00  
48   2024-10-22  06:35:31  01:30:00  
52   2024-10-22  21:59:30  03:00:00  
53   2024-10-22  22:00:25  01:45:00  
63   2024-04-11  23:16:22  01:00:00  
64   2024-04-11  23:17:12  01:30:00  
65   2024-04-11  23:17:59  01:00:00  
70   2024-05-11  21:44:58  01:00:00  
71   2024-05-11  21:45:55  02:00:00  
90   2024-11-20  11:00:16  02:15:00  
96   2024-11-25  23:11:30  04:00:00  
98   2024-11-26  12:54:16  01:27:00  
99   2024-11-26  12:55:05  02:45:00  
113  2024-01-12  23:11:20  02:00:00  
116  2024-06-12  23:12:01  03:00:00  
117  2024-10-12  23:05:51  01:40:00  
118  2024-10-12  23:06:45  01:00:00  
119  2024-10-12  23:07:38  02:05:00  
121  2024-11-12  21:39:14  01:18:00  
122  2024-11-12  21:40:00  01:40:00

In [44]:
diff_df2 = original_df[['Name', 'Subject', 'Start Time', 'End Time', 'RPE', 'Duration', 'time']]    #Jui

In [45]:
diff_df2 = original_df[original_df['Name']== 'Jui']

In [46]:
diff_df2

Subject          Start Time            End Time  RPE Name       dates  \
4    Chemistry 2025-01-15 16:30:00 2025-01-15 17:50:00    3  Jui  2024-10-16   
8    Chemistry 2025-01-15 08:13:00 2025-01-15 09:15:00    3  Jui  2024-10-17   
9     Maths-II 2025-01-15 06:59:00 2025-01-15 07:29:00    3  Jui  2024-10-17   
17     Physics 2025-01-15 17:15:00 2025-01-15 18:45:00    3  Jui  2024-10-18   
18   Chemistry 2025-01-15 19:10:00 2025-01-15 20:00:00    3  Jui  2024-10-18   
..         ...                 ...                 ...  ...  ...         ...   
110    Physics 2025-01-15 17:00:00 2025-01-15 18:30:00    4  Jui  2024-11-29   
111    Physics 2025-01-15 20:34:00 2025-01-15 21:44:00    3  Jui  2024-11-29   
114    Physics 2025-01-15 11:00:00 2025-01-15 12:30:00    3  Jui  2024-04-12   
115    Maths-I 2025-01-15 16:00:00 2025-01-15 17:30:00    2  Jui  2024-04-12   
120  Chemistry 2025-01-15 16:37:00 2025-01-15 17:45:00    3  Jui  2024-11-12   

         time  Duration  
4    17:50:31  01:20:00  
8    09:41:42  01:02:00  
9    19:29:49  00:30:00  
17   07:24:40  01:30:00  
18   07:25:31  00:50:00  
..        ...       ...  
110  19:05:03  01:30:00  
111  21:45:18  01:10:00  
114  21:01:35  01:30:00  
115  21:02:11  01:30:00  
120  18:17:31  01:08:00  

[75 rows x 8 columns]

In [47]:
nupur_df = diff_df.drop(['Start Time', 'End Time', 'dates', 'time', 'RPE', 'Name'], axis=1) 


In [48]:
nupur_df

Subject  Duration
0      Physics  00:58:00
1      Physics  02:05:00
10     Physics  02:30:00
11     Physics  02:15:00
12     English  02:05:00
13     English  01:15:00
27     Biology  01:25:00
28     Biology  01:00:00
49    Maths-II  01:00:00
50     Physics  00:30:00
51    Maths-II  01:10:00
62     Biology  01:30:00
94     Biology  00:45:00
95   Chemistry  01:00:00
100    Physics  01:20:00
108    Maths-I  01:20:00
112    Biology  01:10:00

In [49]:
nupur_df['Duration'] = nupur_df['Duration'].apply(lambda x : pd.to_datetime(str(x)))   

In [50]:
nupur_df

Subject            Duration
0      Physics 2025-01-15 00:58:00
1      Physics 2025-01-15 02:05:00
10     Physics 2025-01-15 02:30:00
11     Physics 2025-01-15 02:15:00
12     English 2025-01-15 02:05:00
13     English 2025-01-15 01:15:00
27     Biology 2025-01-15 01:25:00
28     Biology 2025-01-15 01:00:00
49    Maths-II 2025-01-15 01:00:00
50     Physics 2025-01-15 00:30:00
51    Maths-II 2025-01-15 01:10:00
62     Biology 2025-01-15 01:30:00
94     Biology 2025-01-15 00:45:00
95   Chemistry 2025-01-15 01:00:00
100    Physics 2025-01-15 01:20:00
108    Maths-I 2025-01-15 01:20:00
112    Biology 2025-01-15 01:10:00

In [51]:
nupur_df['Hours'] = nupur_df['Duration'].dt.hour
nupur_df['Minutes'] = nupur_df['Duration'].dt.minute

In [52]:
nupur_df

Subject            Duration  Hours  Minutes
0      Physics 2025-01-15 00:58:00      0       58
1      Physics 2025-01-15 02:05:00      2        5
10     Physics 2025-01-15 02:30:00      2       30
11     Physics 2025-01-15 02:15:00      2       15
12     English 2025-01-15 02:05:00      2        5
13     English 2025-01-15 01:15:00      1       15
27     Biology 2025-01-15 01:25:00      1       25
28     Biology 2025-01-15 01:00:00      1        0
49    Maths-II 2025-01-15 01:00:00      1        0
50     Physics 2025-01-15 00:30:00      0       30
51    Maths-II 2025-01-15 01:10:00      1       10
62     Biology 2025-01-15 01:30:00      1       30
94     Biology 2025-01-15 00:45:00      0       45
95   Chemistry 2025-01-15 01:00:00      1        0
100    Physics 2025-01-15 01:20:00      1       20
108    Maths-I 2025-01-15 01:20:00      1       20
112    Biology 2025-01-15 01:10:00      1       10

In [53]:
nupur_df['Duration'] = nupur_df['Minutes']+ nupur_df['Hours']*60

In [54]:
nupur_df

Subject  Duration  Hours  Minutes
0      Physics        58      0       58
1      Physics       125      2        5
10     Physics       150      2       30
11     Physics       135      2       15
12     English       125      2        5
13     English        75      1       15
27     Biology        85      1       25
28     Biology        60      1        0
49    Maths-II        60      1        0
50     Physics        30      0       30
51    Maths-II        70      1       10
62     Biology        90      1       30
94     Biology        45      0       45
95   Chemistry        60      1        0
100    Physics        80      1       20
108    Maths-I        80      1       20
112    Biology        70      1       10

In [55]:
nupur_df = nupur_df.groupby('Subject').mean()   

In [56]:
nupur_df = nupur_df.drop(['Hours', 'Minutes'], axis=1)

In [57]:
nupur_df

Duration
Subject              
Biology     70.000000
Chemistry   60.000000
English    100.000000
Maths-I     80.000000
Maths-II    65.000000
Physics     96.333333

In [58]:
nupur_df1 = diff_df.drop(['Name', 'dates', 'Start Time', 'End Time', 'Name', 'Duration', 'time'], axis=1)

In [59]:
nupur_df1

Subject  RPE
0      Physics    2
1      Physics    4
10     Physics    3
11     Physics    4
12     English    2
13     English    1
27     Biology    2
28     Biology    1
49    Maths-II    2
50     Physics    1
51    Maths-II    3
62     Biology    1
94     Biology    2
95   Chemistry    2
100    Physics    3
108    Maths-I    3
112    Biology    3

In [60]:
 nupur_df1 = nupur_df1.groupby('Subject').mean()

In [61]:
nupur_df1

RPE
Subject            
Biology    1.800000
Chemistry  2.000000
English    1.500000
Maths-I    3.000000
Maths-II   2.500000
Physics    2.833333

In [62]:
swara_df = diff_df1.drop(['Start Time', 'End Time', 'dates', 'time', 'RPE', 'Name'], axis=1) 


In [63]:
swara_df

Subject  Duration
2    Chemistry  00:32:00
3      Physics  01:00:00
5    Chemistry  01:30:00
6      Physics  01:00:00
7    Chemistry  01:55:00
14     English  02:00:00
15     English  01:15:00
16     English  02:07:00
42     Physics  01:00:00
45     Physics  02:00:00
46   Chemistry  02:45:00
47   Chemistry  01:00:00
48   Chemistry  01:30:00
52     Maths-I  03:00:00
53    Maths-II  01:45:00
63     Maths-I  01:00:00
64     Physics  01:30:00
65    Maths-II  01:00:00
70    Maths-II  01:00:00
71   Chemistry  02:00:00
90   Chemistry  02:15:00
96     Maths-I  04:00:00
98     Physics  01:27:00
99     Physics  02:45:00
113    Physics  02:00:00
116   Maths-II  03:00:00
117   Maths-II  01:40:00
118    Maths-I  01:00:00
119    Physics  02:05:00
121  Chemistry  01:18:00
122  Chemistry  01:40:00

In [64]:
swara_df['Duration'] = swara_df['Duration'].apply(lambda x : pd.to_datetime(str(x))) 
swara_df

Subject            Duration
2    Chemistry 2025-01-15 00:32:00
3      Physics 2025-01-15 01:00:00
5    Chemistry 2025-01-15 01:30:00
6      Physics 2025-01-15 01:00:00
7    Chemistry 2025-01-15 01:55:00
14     English 2025-01-15 02:00:00
15     English 2025-01-15 01:15:00
16     English 2025-01-15 02:07:00
42     Physics 2025-01-15 01:00:00
45     Physics 2025-01-15 02:00:00
46   Chemistry 2025-01-15 02:45:00
47   Chemistry 2025-01-15 01:00:00
48   Chemistry 2025-01-15 01:30:00
52     Maths-I 2025-01-15 03:00:00
53    Maths-II 2025-01-15 01:45:00
63     Maths-I 2025-01-15 01:00:00
64     Physics 2025-01-15 01:30:00
65    Maths-II 2025-01-15 01:00:00
70    Maths-II 2025-01-15 01:00:00
71   Chemistry 2025-01-15 02:00:00
90   Chemistry 2025-01-15 02:15:00
96     Maths-I 2025-01-15 04:00:00
98     Physics 2025-01-15 01:27:00
99     Physics 2025-01-15 02:45:00
113    Physics 2025-01-15 02:00:00
116   Maths-II 2025-01-15 03:00:00
117   Maths-II 2025-01-15 01:40:00
118    Maths-I 2025-01-15 01:00:00
119    Physics 2025-01-15 02:05:00
121  Chemistry 2025-01-15 01:18:00
122  Chemistry 2025-01-15 01:40:00

In [65]:
swara_df['Hours'] = swara_df['Duration'].dt.hour
swara_df['Minutes'] = swara_df['Duration'].dt.minute

In [66]:
swara_df

Subject            Duration  Hours  Minutes
2    Chemistry 2025-01-15 00:32:00      0       32
3      Physics 2025-01-15 01:00:00      1        0
5    Chemistry 2025-01-15 01:30:00      1       30
6      Physics 2025-01-15 01:00:00      1        0
7    Chemistry 2025-01-15 01:55:00      1       55
14     English 2025-01-15 02:00:00      2        0
15     English 2025-01-15 01:15:00      1       15
16     English 2025-01-15 02:07:00      2        7
42     Physics 2025-01-15 01:00:00      1        0
45     Physics 2025-01-15 02:00:00      2        0
46   Chemistry 2025-01-15 02:45:00      2       45
47   Chemistry 2025-01-15 01:00:00      1        0
48   Chemistry 2025-01-15 01:30:00      1       30
52     Maths-I 2025-01-15 03:00:00      3        0
53    Maths-II 2025-01-15 01:45:00      1       45
63     Maths-I 2025-01-15 01:00:00      1        0
64     Physics 2025-01-15 01:30:00      1       30
65    Maths-II 2025-01-15 01:00:00      1        0
70    Maths-II 2025-01-15 01:00:00      1        0
71   Chemistry 2025-01-15 02:00:00      2        0
90   Chemistry 2025-01-15 02:15:00      2       15
96     Maths-I 2025-01-15 04:00:00      4        0
98     Physics 2025-01-15 01:27:00      1       27
99     Physics 2025-01-15 02:45:00      2       45
113    Physics 2025-01-15 02:00:00      2        0
116   Maths-II 2025-01-15 03:00:00      3        0
117   Maths-II 2025-01-15 01:40:00      1       40
118    Maths-I 2025-01-15 01:00:00      1        0
119    Physics 2025-01-15 02:05:00      2        5
121  Chemistry 2025-01-15 01:18:00      1       18
122  Chemistry 2025-01-15 01:40:00      1       40

In [67]:
swara_df['Duration'] = swara_df['Minutes']+ swara_df['Hours']*60

In [68]:
swara_df

Subject  Duration  Hours  Minutes
2    Chemistry        32      0       32
3      Physics        60      1        0
5    Chemistry        90      1       30
6      Physics        60      1        0
7    Chemistry       115      1       55
14     English       120      2        0
15     English        75      1       15
16     English       127      2        7
42     Physics        60      1        0
45     Physics       120      2        0
46   Chemistry       165      2       45
47   Chemistry        60      1        0
48   Chemistry        90      1       30
52     Maths-I       180      3        0
53    Maths-II       105      1       45
63     Maths-I        60      1        0
64     Physics        90      1       30
65    Maths-II        60      1        0
70    Maths-II        60      1        0
71   Chemistry       120      2        0
90   Chemistry       135      2       15
96     Maths-I       240      4        0
98     Physics        87      1       27
99     Physics       165      2       45
113    Physics       120      2        0
116   Maths-II       180      3        0
117   Maths-II       100      1       40
118    Maths-I        60      1        0
119    Physics       125      2        5
121  Chemistry        78      1       18
122  Chemistry       100      1       40

In [69]:
swara_df = swara_df.drop(['Hours', 'Minutes'], axis=1)

In [70]:
swara_df.groupby('Subject').mean()

Duration
Subject              
Chemistry   98.500000
English    107.333333
Maths-I    135.000000
Maths-II   101.000000
Physics     98.555556

In [71]:
jui_df = diff_df2.drop(['Start Time', 'End Time', 'dates', 'time', 'RPE', 'Name'], axis=1) 

In [72]:
jui_df['Duration'] = jui_df['Duration'].apply(lambda x : pd.to_datetime(str(x))) 

In [73]:
jui_df['Hours'] = jui_df['Duration'].dt.hour
jui_df['Minutes'] = jui_df['Duration'].dt.minute

In [74]:
jui_df['Duration'] = jui_df['Minutes']+ jui_df['Hours']*60

In [75]:
jui_df = jui_df.drop(['Hours', 'Minutes'], axis=1)

In [76]:
jui_df.groupby('Subject').mean()

Duration
Subject             
Chemistry  79.148148
English    97.500000
Maths-I    74.583333
Maths-II   79.666667
Physics    75.090909

In [77]:
import matplotlib.pyplot as plt

In [78]:
comdf = pd.read_csv('combinedData - Sheet1-2.csv')
comdf

name   subject    duration 
0   nupur    Biology   70.000000
1   nupur  Chemistry   60.000000
2   nupur    English  100.000000
3   nupur    Maths-I   80.000000
4   nupur   Maths-II   65.000000
5   nupur    Physics   96.333333
6     jui    Biology    0.000000
7     jui  Chemistry   79.150000
8     jui    English   97.500000
9     jui    Maths-I   74.580000
10    jui   Maths-II   79.660000
11    jui    Physics   75.090000
12  swara    Biology    0.000000
13  swara  Chemistry   98.500000
14  swara    English  107.330000
15  swara    Maths-I  135.000000
16  swara   Maths-II  101.000000
17  swara    Physics   98.550000

## plotting graphs for students investing time for the studies subjectwise

In [79]:
new_df1

Duration
Name             
Jui     77.800000
Nupur   82.235294
Swara  104.483871

In [80]:
new_df2

Duration     Hours    Minutes
Subject                                   
Biology     70.000000  0.800000  22.000000
Chemistry   83.736842  1.000000  23.736842
English    102.428571  1.428571  16.714286
Maths-I     89.117647  1.235294  15.000000
Maths-II    83.736842  1.052632  20.578947
Physics     84.243243  1.000000  24.243243

In [81]:
new_df2['Duration'] =new_df2['Duration'].apply(lambda x : pd.to_datetime(str(x))) 

ValueError: Given date string "70.0" not likely a datetime, at position 0

In [ ]:
new_df2['Hours'] = new_df2['Duration'].dt.hour
new_df2['Minutes'] = new_df2['Duration'].dt.minute

In [ ]:
new_df2['Duration'] = new_df2['Minutes']+ new_df2['Hours']*60

In [ ]:
new_df2

In [ ]:
nupur_df

In [ ]:
# Pivot the DataFrame to reshape it
pivot_df = comdf.pivot(index='name', columns='subject ', values='duration ')

# Plot the grouped bar chart
ax = pivot_df.plot(kind='bar', figsize=(12, 7), width=0.8)


# Customize the chart
plt.title('Subject Duration by Name', fontsize=16)
plt.xlabel('Name', fontsize=14)
plt.ylabel('Duration', fontsize=14)
plt.xticks(rotation=0, fontsize=12)
plt.legend(title='Subjects', fontsize=12)
plt.tight_layout()
for container in ax.containers:
    ax.bar_label(container, fmt='%.1f', label_type='edge', fontsize=10)

# Show the plot
plt.show()


## Plot graphs for students working specifically at dawn, morning, afternoon and evening


In [ ]:
# Extract the hour from the start_time
original_df['start_hour'] = original_df['Start Time'].dt.hour
original_df['Duration'] = (original_df['End Time'] - original_df['Start Time']).dt.total_seconds() / 3600

In [ ]:
def classify_time(hour):
    if 0 <= hour < 6:
        return 'Dawn'
    elif 6 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 18:
        return 'Afternoon'
    else:
        return 'Evening'


In [ ]:
original_df['time_slot'] = original_df['start_hour'].apply(classify_time)

In [ ]:
original_df

In [ ]:
pivot_df = original_df.pivot_table(index='Name', columns='time_slot', values='Duration', aggfunc='sum', fill_value=0)

# Plot the grouped bar chart
ax = pivot_df.plot(kind='bar', figsize=(12, 7), width=0.8)

# Customize the chart
plt.title('Study Duration by Time Slot for Each Student', fontsize=16)
plt.xlabel('Name', fontsize=14)
plt.ylabel('Duration (hours)', fontsize=14)
plt.xticks(rotation=0, fontsize=12)
plt.legend(title='Time Slot', fontsize=12)
plt.tight_layout()

# Add labels to the bars
for container in ax.containers:
    ax.bar_label(container, fmt='%.1f', label_type='edge', fontsize=10)

# Show the plot
plt.show()

In [ ]:
pivot_df

In [ ]:
pivot_df.plot(kind='line', figsize=(12, 7), marker='o')

# Customize the chart
plt.title('Study Duration by Time Slot for Each Student', fontsize=16)
plt.xlabel('Name', fontsize=14)
plt.ylabel('Duration (hours)', fontsize=14)
plt.xticks(rotation=0, fontsize=12)
plt.legend(title='Time Slot', fontsize=12)
plt.tight_layout()

# Show the plot
plt.show()

In [ ]:
# Group by student name and time slot, and calculate the average duration spent in each time slot
pivot_df = original_df.groupby(['Name', 'time_slot'])['Duration'].mean().unstack(fill_value=0)

time_slot_colors = {
    'Dawn': 'blue',
    'Morning': 'green',
    'Afternoon': 'orange',
    'Evening': 'red'}

for time_slot in pivot_df.columns:
    plt.plot(pivot_df.index, pivot_df[time_slot], marker='o', label=time_slot, color=time_slot_colors[time_slot])

plt.title('Average Study Duration by Time Slot for Each Student', fontsize=16)
plt.xlabel('Name', fontsize=14)
plt.ylabel('Duration', fontsize=14)
plt.xticks(rotation=0, fontsize=12)
plt.yticks(rotation=0, fontsize=12)
plt.legend(title='Time Slot', fontsize=12)


# Show the plot
plt.tight_layout()
plt.show()


In [ ]:
ddf= pivot_df.transpose()
ddf

In [ ]:
plt.plot(ddf)


plt.xlabel('time_slot', fontsize=14)
plt.ylabel('Duration', fontsize=14)
plt.legend(title='Time Slot', fontsize=12)
Name_colors = {
    'Jui': 'blue',
    'Swara': 'green',
    'Nupur': 'orange'}
for Name in ddf.columns:
    plt.plot(ddf.index, ddf[Name], marker='o', label=Name, color=Name_colors[Name])

plt.show()


In [ ]:
diff_df1['start_hour'] = diff_df1['Start Time'].dt.hour
diff_df1['Duration'] = (diff_df1['End Time'] - diff_df1['Start Time']).dt.total_seconds() / 3600
diff_df1['time_slot'] = diff_df1['start_hour'].apply(classify_time)



In [ ]:
aggregated_df1 = diff_df1.groupby('time_slot').agg({
    'Duration': 'mean',  
    'RPE': 'mean'       
}).reset_index()

fig, ax = plt.subplots()

# Plot Duration
aggregated_df1.plot.bar(x='time_slot', y='Duration', position=0, ax=ax, width =0.2, label='Duration', color='blue')

# Plot RPE
aggregated_df1.plot.bar(x='time_slot', y='RPE', ax=ax, position=1, width =0.2, label='RPE', color='green')


In [ ]:
fig, ax = plt.subplots(figsize=(10, 6))     #nupur

# Bar width and positions
bar_width = 0.35
x = np.arange(len(aggregated_df1))

# Plot the max bars in the background
ax.bar(x - bar_width / 2, [3] * len(aggregated_df1), width=bar_width, color='gray', alpha=0.3, label='Max Duration')
ax.bar(x + bar_width / 2, [5] * len(aggregated_df1), width=bar_width, color='lightblue', alpha=0.3, label='Max RPE')
# Plot the individual duration and RPE bars
duration_bars = ax.bar(x - bar_width / 2, aggregated_df1['Duration'], width=bar_width, color='orange', label='Duration')
rpe_bars = ax.bar(x + bar_width / 2, aggregated_df1['RPE'], width=bar_width, color='blue', label='RPE')

ax.set_xlabel('time_slot')  # Changed x-axis label
ax.set_ylabel('Duration')  # Updated y-axis label
ax.set_title('Joint Bar Graph of Duration and RPE')
ax.set_xticks(x)
ax.set_xticklabels(['Dawn', 'Morning', 'Afternoon', 'Evening'])  # Updated x-axis tick labels
ax.set_ylim(0, 5)  # Adjusted y-axis scale
ax.legend()

# Show the plot
plt.tight_layout()
plt.show()


In [ ]:
x

In [ ]:
diff_df2['start_hour'] = diff_df2['Start Time'].dt.hour
diff_df2['Duration'] = (diff_df2['End Time'] - diff_df2['Start Time']).dt.total_seconds() / 3600
diff_df2['time_slot'] = diff_df2['start_hour'].apply(classify_time)


In [ ]:
aggregated_df2 = diff_df2.groupby('time_slot').agg({
    'Duration': 'mean',  # Summing up durations
    'RPE': 'mean'       # Taking the average RPE        #swara
}).reset_index()

fig, ax = plt.subplots()

# Plot Duration
aggregated_df2.plot.bar(x='time_slot', y='Duration', position=0, ax=ax, width =0.2, label='Duration', color='black')

# Plot RPE
aggregated_df2.plot.bar(x='time_slot', y='RPE', ax=ax, position=1, width =0.2, label='RPE', color='grey')


In [ ]:
fig, ax = plt.subplots(figsize=(10, 6))

# Bar width and positions
bar_width = 0.35
x = np.arange(len(aggregated_df2))

# Plot the max bars in the background
ax.bar(x - bar_width / 2, [3] * len(aggregated_df2), width=bar_width, color='gray', alpha=0.3, label='Max Duration')
ax.bar(x + bar_width / 2, [5] * len(aggregated_df2), width=bar_width, color='lightblue', alpha=0.3, label='Max RPE')
# Plot the individual duration and RPE bars
duration_bars = ax.bar(x - bar_width / 2, aggregated_df2['Duration'], width=bar_width, color='orange', label='Duration')
rpe_bars = ax.bar(x + bar_width / 2, aggregated_df2['RPE'], width=bar_width, color='blue', label='RPE')

ax.set_xlabel('time_slot')  # Changed x-axis label
ax.set_ylabel('Duration')  # Updated y-axis label
ax.set_title('Joint Bar Graph of Duration and RPE')
ax.set_xticks(x)
ax.set_xticklabels(['Dawn', 'Morning', 'Afternoon', 'Evening'])  # Updated x-axis tick labels
ax.set_ylim(0, 5)  # Adjusted y-axis scale
ax.legend()

# Show the plot
plt.tight_layout()
plt.show()


In [ ]:
diff_df['start_hour'] = diff_df['Start Time'].dt.hour
diff_df['Duration'] = (diff_df['End Time'] - diff_df['Start Time']).dt.total_seconds() / 3600
diff_df['time_slot'] = diff_df['start_hour'].apply(classify_time)



In [ ]:
diff_df

In [ ]:
aggregated_df3 = diff_df.groupby('time_slot').agg({
    'Duration': 'mean',  # Summing up durations
    'RPE': 'mean'       # Taking the average RPE        #swara
}).reset_index()


In [ ]:
x = np.arange(len(aggregated_df3))  # Make sure this matches the length of the bars

fig, ax = plt.subplots(figsize=(10, 6))

# Bar width and positions
bar_width = 0.35

# Plot the max bars in the background
ax.bar(x - bar_width / 2, [3] * len(aggregated_df3), width=bar_width, color='gray', alpha=0.3, label='Max Duration')
ax.bar(x + bar_width / 2, [5] * len(aggregated_df3), width=bar_width, color='lightblue', alpha=0.3, label='Max RPE')

# Plot the individual duration and RPE bars
duration_bars = ax.bar(x - bar_width / 2, aggregated_df3['Duration'], width=bar_width, color='orange', label='Duration')
rpe_bars = ax.bar(x + bar_width / 2, aggregated_df3['RPE'], width=bar_width, color='blue', label='RPE')

# Setting labels and titles
ax.set_xlabel('time_slot')
ax.set_ylabel('Duration')
ax.set_title('Joint Bar Graph of Duration and RPE')

# Adjust x-ticks to correspond with the length of the x array
ax.set_xticks(x)
ax.set_xticklabels([ 'Morning', 'Afternoon', 'Evening'])

# Set y-axis limits
ax.set_ylim(0, 5)

# Display the legend
ax.legend()

# Show the plot
plt.tight_layout()
plt.show()

In [ ]:
diff_df

In [ ]:
aggregated_sub1 = diff_df.groupby('Subject').agg({
    'Duration': 'mean',  # Summing up durations
    'RPE': 'mean'       # Taking the average RPE        #swara
}).reset_index()

In [ ]:

# Specify the subject you want to plot (e.g., "Math")
subject_name = 'Chemistry'

# Filter the DataFrame for the selected subject
subject_data = diff_df[diff_df['Subject'] == subject_name]

# Ensure that the time slots are sorted correctly, or you might need to do it manually if not sorted
time_slots = ['Dawn', 'Morning', 'Afternoon', 'Evening']

# Map the time slots to the data
subject_data_sorted = subject_data.set_index('time_slot').reindex(time_slots).reset_index()

# Bar width and positions
bar_width = 0.2  # Narrow bar width
x = np.arange(len(time_slots))  # Time slots positions

# Create a figure and axis for the plot
fig, ax = plt.subplots(figsize=(10, 6))

# Plot the max bars in the background
ax.bar(x - bar_width / 2, [3] * len(subject_data_sorted), width=bar_width, color='gray', alpha=0.3, label='Max Duration')
ax.bar(x + bar_width / 2, [5] * len(subject_data_sorted), width=bar_width, color='lightblue', alpha=0.3, label='Max RPE')

# Plot the individual duration and RPE bars
ax.bar(x - bar_width / 2, subject_data_sorted['Duration'], width=bar_width, color='orange', label='Duration')
ax.bar(x + bar_width / 2, subject_data_sorted['RPE'], width=bar_width, color='blue', label='RPE')

# Set labels and title
ax.set_xlabel('Time Slot')
ax.set_ylabel('Duration')
ax.set_title(f'Joint Bar Graph of Duration and RPE for {subject_name}')

# Set x-ticks and labels for time slots
ax.set_xticks(x)
ax.set_xticklabels(time_slots)

# Set y-axis limit for better visibility
ax.set_ylim(0, 5)  # Adjusted y-axis scale

# Add a legend
ax.legend()

# Show the plot
plt.tight_layout()
plt.show()


In [ ]:

# Specify the subject you want to plot (e.g., "Math")
subject_name = 'Physics'

# Filter the DataFrame for the selected subject
subject_data = diff_df[diff_df['Subject'] == subject_name]

# Ensure that the time slots are sorted correctly, or you might need to do it manually if not sorted
time_slots = ['Dawn', 'Morning', 'Afternoon', 'Evening']

# Group by time_slot and aggregate the values (e.g., taking the mean)
subject_data_grouped = subject_data.groupby('time_slot').agg({'Duration': 'mean', 'RPE': 'mean'}).reset_index()

# Reorder the data to match the desired time slots order
subject_data_sorted = subject_data_grouped.set_index('time_slot').reindex(time_slots).reset_index()

# Bar width and positions
bar_width = 0.2  # Narrow bar width
x = np.arange(len(time_slots))  # Time slots positions

# Create a figure and axis for the plot
fig, ax = plt.subplots(figsize=(10, 6))

# Plot the max bars in the background
ax.bar(x - bar_width / 2, [3] * len(subject_data_sorted), width=bar_width, color='gray', alpha=0.3, label='Max Duration')
ax.bar(x + bar_width / 2, [5] * len(subject_data_sorted), width=bar_width, color='lightblue', alpha=0.3, label='Max RPE')

# Plot the individual duration and RPE bars
ax.bar(x - bar_width / 2, subject_data_sorted['Duration'], width=bar_width, color='orange', label='Duration')
ax.bar(x + bar_width / 2, subject_data_sorted['RPE'], width=bar_width, color='blue', label='RPE')

# Set labels and title
ax.set_xlabel('Time Slot')
ax.set_ylabel('Duration')
ax.set_title(f'Joint Bar Graph of Duration and RPE for {subject_name}')

# Set x-ticks and labels for time slots
ax.set_xticks(x)
ax.set_xticklabels(time_slots)

# Set y-axis limit for better visibility
ax.set_ylim(0, 5)  # Adjusted y-axis scale

# Add a legend
ax.legend()

# Show the plot
plt.tight_layout()
plt.show()


In [ ]:
nupur_df

## number of sessions logged per subject for every girl

In [ ]:
diff_df

In [ ]:
diff_df.describe()

In [ ]:
diff_df = diff_df.drop(['Start Time', 'End Time', 'Name', 'dates', 'time'], axis=1)

In [ ]:
Nupur_phy_row = diff_df.loc[diff_df['Subject'] == 'Physics']

In [ ]:
Nupur_phy_row['Duration'] =Nupur_phy_row['Duration'].apply(lambda x : pd.to_datetime(str(x))) 

In [ ]:
Nupur_phy_row['Hours'] = Nupur_phy_row['Duration'].dt.hour
Nupur_phy_row['Minutes'] = Nupur_phy_row['Duration'].dt.minute

In [ ]:
Nupur_phy_row['Duration'] = Nupur_phy_row['Minutes']+ Nupur_phy_row['Hours']*60

In [ ]:
Nupur_phy_row

In [ ]:
Nupur_phy_row.drop(['Hours', 'Minutes'], axis=1)

In [ ]:
# creating separte df for each subjcet to get plot 
Nupur_phy_row =diff_df[diff_df['Subject']== 'Physics']
Nupur_maths1_row =diff_df[diff_df['Subject']== 'Maths-I']
Nupur_maths2_row =diff_df[diff_df['Subject']== 'Maths-II']
Nupur_chem_row= diff_df[diff_df['Subject']== 'Chemistry']
Nupur_bio_row =diff_df[diff_df['Subject']== 'Biology']
Nupur_eng_row =diff_df[diff_df['Subject']== 'English']

In [ ]:
Nupur_maths1_row

In [ ]:
Nupur_maths2_row

In [ ]:
Nupur_phy_row

In [ ]:
Nupur_chem_row

In [ ]:
Nupur_eng_row

# Jui

In [ ]:
diff_df2

In [ ]:
# creating separte df for each subjcet to get plot 
Jui_phy_row =diff_df2[diff_df2['Subject']== 'Physics']
Jui_maths1_row =diff_df2[diff_df2['Subject']== 'Maths-I']
Jui_maths2_row =diff_df2[diff_df2['Subject']== 'Maths-II']
Jui_chem_row= diff_df2[diff_df2['Subject']== 'Chemistry']
Jui_eng_row =diff_df2[diff_df2['Subject']== 'English']

In [ ]:
Jui_phy_row['Duration'] =Jui_phy_row['Duration'].apply(lambda x : pd.to_datetime(str(x))) 
Jui_phy_row['Hours'] = Jui_phy_row['Duration'].dt.hour
Jui_phy_row['Minutes'] = Jui_phy_row['Duration'].dt.minute
Jui_phy_row['Duration'] = Jui_phy_row['Minutes']+ Jui_phy_row['Hours']*60

Jui_maths1_row['Duration'] =Jui_maths1_row['Duration'].apply(lambda x : pd.to_datetime(str(x))) 
Jui_maths1_row['Hours'] = Jui_maths1_row['Duration'].dt.hour
Jui_maths1_row['Minutes'] = Jui_maths1_row['Duration'].dt.minute
Jui_maths1_row['Duration'] = Jui_maths1_row['Minutes']+ Jui_maths1_row['Hours']*60

Jui_maths2_row['Duration'] =Jui_maths2_row['Duration'].apply(lambda x : pd.to_datetime(str(x))) 
Jui_maths2_row['Hours'] = Jui_maths2_row['Duration'].dt.hour
Jui_maths2_row['Minutes'] = Jui_maths2_row['Duration'].dt.minute
Jui_maths2_row['Duration'] = Jui_maths2_row['Minutes']+ Jui_maths2_row['Hours']*60

Jui_chem_row['Duration'] =Jui_chem_row['Duration'].apply(lambda x : pd.to_datetime(str(x))) 
Jui_chem_row['Hours'] = Jui_chem_row['Duration'].dt.hour
Jui_chem_row['Minutes'] = Jui_chem_row['Duration'].dt.minute
Jui_chem_row['Duration'] = Jui_chem_row['Minutes']+ Jui_chem_row['Hours']*60

Jui_eng_row['Duration'] =Jui_eng_row['Duration'].apply(lambda x : pd.to_datetime(str(x))) 
Jui_eng_row['Hours'] = Jui_eng_row['Duration'].dt.hour
Jui_eng_row['Minutes'] = Jui_eng_row['Duration'].dt.minute
Jui_eng_row['Duration'] = Jui_eng_row['Minutes']+ Jui_eng_row['Hours']*60



In [ ]:
def create_subject_summary(df):
    # Group by subject and calculate metrics
    summary = df.groupby('Subject').agg({
        'RPE': ['count', 'mean'],
        'Duration': 'mean'
    })
    
    # Flatten column names and rename
    summary.columns = ['Sessions', 'Avg RPE', 'Avg Dura']
    
    # Reset index to make Subject a column
    summary = summary.reset_index()
    
    # Round the averages to 2 decimal places
    summary['Avg RPE'] = summary['Avg RPE'].round(2)
    summary['Avg Dura'] = summary['Avg Dura'].round(2)
    
    return summary

In [ ]:
jui_summary = create_subject_summary(diff_df2)

In [ ]:
jui_summary

def plot_subject_summary(summary_df):
    # Create figure with secondary y-axis
    fig = go.Figure()
    
    # Add bars for Sessions
    fig.add_trace(
        go.Bar(name='Sessions', 
               x=summary_df['Subject'], 
               y=summary_df['Sessions'],
               marker_color='rgb(55, 83, 109)',
               text=summary_df['Sessions'],
               textposition='auto')
    )
    
    # Add bars for Average RPE
    fig.add_trace(
        go.Bar(name='Average RPE', 
               x=summary_df['Subject'], 
               y=summary_df['Avg RPE'],
               marker_color='rgb(26, 118, 255)',
               text=summary_df['Avg RPE'].round(2),
               textposition='auto')
    )
    
    # Add bars for Average Duration
    fig.add_trace(
        go.Bar(name='Average Duration (min)', 
               x=summary_df['Subject'], 
               y=summary_df['Avg Duration'],
               marker_color='rgb(158, 202, 225)',
               text=summary_df['Avg Duration'].round(1),
               textposition='auto')
    )

    # Update layout
    fig.update_layout(
        title='Subject-wise Analysis',
        title_x=0,
        barmode='group',
        width=1000,
        height=600,
        yaxis_title='Values',
        xaxis_title='Subjects',
        bargap=0.15,
        bargroupgap=0.1,
        showlegend=True,
        legend=dict(
            yanchor="top",
            y=1.15,
            xanchor="center",
            x=0.5,
            orientation="h"
        )
    )
    
    return fig

In [ ]:
plot_subject_summary(jui_summary)

# Swara

In [ ]:
def create_subject_summary(df):
    # Group by subject and calculate metrics
    summary = df.groupby('Subject').agg({
        'RPE': ['count', 'mean'],
        'Duration': 'mean'
    })
    
    # Flatten column names and rename
    summary.columns = ['Sessions', 'Avg RPE', 'Avg Dura']
    
    # Reset index to make Subject a column
    summary = summary.reset_index()
    
    # Round the averages to 2 decimal places
    summary['Avg RPE'] = summary['Avg RPE'].round(2)
    summary['Avg Dura'] = summary['Avg Dura'].round(2)
    
    return summary

In [ ]:
#diff_df1['Duration'] = pd.to_numeric(diff_df1['Duration'], errors='coerce')
#diff_df1['RPE'] = pd.to_numeric(diff_df1['RPE'], errors='coerce')

In [ ]:
swara_summary = create_subject_summary(diff_df1)

In [ ]:
def create_subject_summary(df):
    # Group by subject and calculate metrics
    summary = df.groupby('Subject').agg({
        'RPE': ['count', 'mean'],
        'Duration': 'mean'
    })
    
    # Flatten column names and rename
    summary.columns = ['Sessions', 'Avg RPE', 'Avg Dura']
    
    # Reset index to make Subject a column
    summary = summary.reset_index()
    
    # Round the averages to 2 decimal places
    summary['Avg RPE'] = summary['Avg RPE'].round(2)
    summary['Avg Dura'] = summary['Avg Dura'].round(2)
    
    return summary

In [ ]:
plot_subject_summary(swara_summary)

# Nupur

In [ ]:
# Assuming 'diff_df' is your DataFrame
# Convert the 'Duration' column to timedelta
diff_df['Duration'] = pd.to_timedelta(diff_df['Duration'])

# Convert timedelta to hours (float)
diff_df['Duration (Hours)'] = diff_df['Duration'].dt.total_seconds() / 3600

# Display the updated DataFrame
print(diff_df)


In [ ]:
diff_df.drop(['Duration'], axis=1)

In [ ]:
def create_subject_summary(df):
    # Group by subject and calculate metrics
    summary = df.groupby('Subject').agg({
        'RPE': ['count', 'mean'],
        'Duration': 'mean'
    })
    
    # Flatten column names and rename
    summary.columns = ['Sessions', 'Avg RPE', 'Avg Dura']
    
    # Reset index to make Subject a column
    summary = summary.reset_index()
    
    # Round the averages to 2 decimal places
    summary['Avg RPE'] = summary['Avg RPE'].round(2)
    summary['Avg Dura'] = summary['Avg Dura'].round(2)
    
    return summary

In [ ]:
nupur_summary = create_subject_summary(diff_df)

In [ ]:
plot_subject_summary(nupur_summary)

# combined graph for all there girls

In [ ]:
# Create separate dataframes for each group
Nupur_data = {
    'Subject': ['Biology', 'Chemistry', 'English', 'Maths-I', 'Maths-II', 'Physics'],
    'Sessions': [5, 1, 2, 1, 2, 6],
    'RPE': [1.8, 2.0, 1.5, 3.0, 2.5, 2.83],
    'Duration': [1.2, 1.0, 1.7, 1.3, 1.1, 1.6]
}

Swara_data = {
    'Subject': ['Chemistry', 'English', 'Maths-I', 'Maths-II', 'Physics'],
    'Sessions': [10, 3, 4, 5, 9],
    'RPE': [2.1, 2.0, 2.0, 2.2, 2.33],
    'Duration': [1.6, 1.8, 2.2, 1.7, 1.6]
}

Jui_data = {
    'Subject': ['Chemistry', 'English', 'Maths-I', 'Maths-II', 'Physics'],
    'Sessions': [27, 2, 12, 12, 22],
    'RPE': [3.04, 2.0, 2.58, 2.58, 3.05],
    'Duration': [1.3, 1.6, 1.2, 1.3, 1.2]
}

# Create figure
fig = go.Figure()

# Add traces for Group 1
fig.add_trace(go.Bar(
    name='Nupur Sessions',
    x=Nupur_data['Subject'],
    y=Nupur_data['Sessions'],
    marker_color='rgb(31, 119, 180)'
))
fig.add_trace(go.Bar(
    name='Nupur RPE',
    x=Nupur_data['Subject'],
    y=Nupur_data['RPE'],
    marker_color='rgb(174, 199, 232)'
))
fig.add_trace(go.Bar(
    name='Nupur Duration',
    x=Nupur_data['Subject'],
    y=Nupur_data['Duration'],
    marker_color='rgb(219, 236, 247)'
))

# Add traces for Group 2
fig.add_trace(go.Bar(
    name='Swara Sessions',
    x=Swara_data['Subject'],
    y=Swara_data['Sessions'],
    marker_color='rgb(44, 160, 44)'
))
fig.add_trace(go.Bar(
    name='Swara RPE',
    x=Swara_data['Subject'],
    y=Swara_data['RPE'],
    marker_color='rgb(152, 223, 138)'
))
fig.add_trace(go.Bar(
    name='Swara Duration',
    x=Swara_data['Subject'],
    y=Swara_data['Duration'],
    marker_color='rgb(219, 245, 219)'
))

# Add traces for Group 3
fig.add_trace(go.Bar(
    name='Jui Sessions',
    x=Jui_data['Subject'],
    y=Jui_data['Sessions'],
    marker_color='rgb(255, 127, 14)'
))
fig.add_trace(go.Bar(
    name='Jui RPE',
    x=Jui_data['Subject'],
    y=Jui_data['RPE'],
    marker_color='rgb(255, 187, 120)'
))
fig.add_trace(go.Bar(
    name='Jui Duration',
    x=Jui_data['Subject'],
    y=Jui_data['Duration'],
    marker_color='rgb(255, 229, 204)'
))

fig.update_layout(
        title='Subject-wise Comparative Analysis',
        title_x=0.5,
        barmode='group',
        width=1100,
        height=600,
        yaxis_title='Values',
        xaxis_title='Subjects',
        bargap=0.15,
        bargroupgap=0.1,
        showlegend=True,
        legend=dict(
            yanchor="top",
            y=1.10,
            xanchor="center",
            x=0.5,
            orientation="h"
        )
    )
fig.update_layout(plot_bgcolor='rgb(245, 245, 250)', paper_bgcolor='rgb(245, 245, 250)')

# Show the figure
fig.show()